File: Sources/CountriesFlagsObjC/source/Countries/UIImage+Algeria.m
//
//  UIImage+Algeria.m
//  CountriesFlags
//

#import "Countries/UIImage+Algeria.h"

#import "Extensions/UIBezierPath+Star.h"
#import "Extensions/UIBezierPath+AffineTransform.h"
#import "Extensions/CGSize+Extension.h"

@implementation UIImage (Algeria)

+ (UIImage * _Nullable)flagAlgeriaWithSize:(CGSize)size {
    UIColor* colorRed = [UIColor colorWithRed:217.0f/255.0f
                                        green:0.0f
                                         blue:18.0f/255.0f
                                        alpha:1.0f];
    UIColor* colorGreen = [UIColor colorWithRed:0.0f
                                          green:102.0f/255.0f
                                           blue:51.0f/255.0f
                                          alpha:1.0f];
    UIColor* colorWhite = [UIColor whiteColor];
    
    UIGraphicsBeginImageContextWithOptions(size, NO, 0.0f);
    
    [colorWhite setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorGreen setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width / 2.0f, size.height));

    [colorRed setFill];
    [[UIBezierPath bezierPathWithArcCenter:CGSizeMidPoint(size)
                                    radius:size.height / 4.0f
                                startAngle:0.0f
                                  endAngle:M_PI * 2.0f
                                 clockwise:YES] fill];

    [colorWhite setFill];
    CGPoint arcCenter = CGPointMake(size.width / 2.0f + size.height / 20.0f * 1.207628020, size.height / 2.0f);
    CGFloat arcRadius = size.height / 20.0f * 8.0f / 2.0f;
    [[UIBezierPath bezierPathWithArcCenter:arcCenter
                                    radius:arcRadius
                                startAngle:0.0f
                                  endAngle:M_PI * 2.0f
                                 clockwise:YES] fill];

    [colorGreen setFill];
    NSUnitAngle* unitDegrees = [NSUnitAngle degrees];
    NSUnitAngle* unitRadians = [NSUnitAngle radians];
    double startAngle = [[[[NSMeasurement alloc] initWithDoubleValue:107.6
                                                                unit:unitDegrees] measurementByConvertingToUnit:unitRadians] doubleValue];
    double endAngle = [[[[NSMeasurement alloc] initWithDoubleValue:-107.6
                                                              unit:unitDegrees] measurementByConvertingToUnit:unitRadians] doubleValue];
    [[UIBezierPath bezierPathWithArcCenter:arcCenter
                                    radius:arcRadius
                                startAngle:startAngle
                                  endAngle:endAngle
                                 clockwise:YES] fill];

    [colorRed setFill];
    CGPoint starCenter = CGPointMake(size.width / 2.0f + size.height / 20.0f * 2.022542486, size.height / 2.0f);
    UIBezierPath* path = [UIBezierPath starAtPoint:starCenter
                                            radius:size.height / 20.0f * 5.0f / 2.0f
                                              type:FlagStarTypeFivePointed];
    [path rotateToAngle:18.0f center:starCenter];
    [path fill];

    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image;
}

@end

File: Sources/CountriesFlags/Countries/UIImage+Algeria.swift
//
//  UIImage+Algeria.swift
//  CountriesFlags
//

// swiftlint:disable function_body_length

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_Algeria "wikipedia"
     
     Generates the flag of Algeria.
     For more information see [wikipedia].

     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of Algeria.
     */
    internal static func flagAlgeria(with size: CGSize) -> UIImage? {
        let colorRed = UIColor(red: 217.0 / 255.0,
                               green: 0.0,
                               blue: 18.0 / 255.0,
                               alpha: 1.0)
        let colorGreen = UIColor(red: 0.0,
                                 green: 102.0 / 255.0,
                                 blue: 51.0 / 255.0,
                                 alpha: 1.0)
        let colorWhite = UIColor.white

        UIGraphicsBeginImageContextWithOptions(size, false, 0.0)

        colorWhite.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorGreen.setFill()
        UIRectFill(CGRect(origin: .zero,
                          size: CGSize(width: size.width / 2.0, height: size.height)))

        colorRed.setFill()
        UIBezierPath(arcCenter: size.midPoint,
                     radius: size.height / 4.0,
                     startAngle: 0.0,
                     endAngle: .pi * 2.0,
                     clockwise: true).fill()

        colorWhite.setFill()
        UIBezierPath(arcCenter: CGPoint(x: size.width / 2.0 + size.height / 20.0 * 1.207628020, y: size.height / 2.0),
                     radius: size.height / 20.0 * 4.0,
                     startAngle: 0.0,
                     endAngle: .pi * 2.0,
                     clockwise: true).fill()

        colorGreen.setFill()
        let startAngle = Measurement(value: 107.6, unit: UnitAngle.degrees).converted(to: .radians).value
        let endAngle = Measurement(value: -107.6, unit: UnitAngle.degrees).converted(to: .radians).value
        UIBezierPath(arcCenter: CGPoint(x: size.width / 2.0 + size.height / 20.0 * 1.207628020, y: size.height / 2.0),
                     radius: size.height / 20.0 * 4.0,
                     startAngle: startAngle,
                     endAngle: endAngle,
                     clockwise: true).fill()

        colorRed.setFill()
        let starCenter = CGPoint(x: size.width / 2.0 + size.height / 20.0 * 2.022542486, y: size.height / 2.0)
        let path = UIBezierPath.star(at: starCenter, radius: size.height / 20.0 * 2.5)
        path.rotate(to: 18.0, around: starCenter)
        path.fill()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image
    }
}

#endif

// swiftlint:enable function_body_length
